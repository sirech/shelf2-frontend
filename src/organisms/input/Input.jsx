// @flow

import React from 'react'
import { connect } from 'react-redux'
import { createStructuredSelector } from 'reselect'
import * as R from 'ramda'

import { validSelector, touchedSelector } from './selectors'

import './styles.css'

import { Control, Errors } from 'react-redux-form'
import { FormGroup, Label, Input as BaseInput, FormFeedback } from 'reactstrap'

type Props = {
  name: string,
  type: string,
  valid: boolean,
  touched: boolean,
  messages: Object,
}

// exported for testing
export class Input extends React.Component {
  props: Props

  isValid() {
    const { valid, touched } = this.props

    if (!touched) {
      return undefined
    }

    return valid
  }

  isInvalid() {
    const { valid, touched } = this.props

    if (!touched) {
      return undefined
    }

    return !valid
  }

  render() {
    const { name, type, messages } = this.props
    const model = `.${name}`

    return (
      <FormGroup>
        <Label for={name}>{name}</Label>
        <Control
          model={model}
          type={type}
          name={name}
          id={name}
          valid={this.isValid()}
          invalid={this.isInvalid()}
          component={BaseInput}
          {...R.omit(['name', 'type', 'messages', 'valid', 'touched'])(
            this.props
          )}
        />
        <Errors
          model={model}
          wrapper={props => props.children}
          component={props => <FormFeedback>{props.children}</FormFeedback>}
          messages={messages}
          show={{ touched: true, valid: false }}
        />
      </FormGroup>
    )
  }
}

export default connect((state, props) =>
  createStructuredSelector({
    valid: validSelector,
    touched: touchedSelector,
  })(state, props)
)(Input)
