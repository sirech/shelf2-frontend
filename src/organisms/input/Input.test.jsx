import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'

import { Input } from './Input'

describe('components', () => {
  describe('Input', () => {
    it('renders correctly', () => {
      const component = shallow(<Input name="title" type="text" />)
      expect(toJson(component)).toMatchSnapshot()
    })

    it('renders correctly for different types', () => {
      const component = shallow(<Input name="year" type="number" />)
      expect(toJson(component)).toMatchSnapshot()
    })

    it('displays error', () => {
      const component = shallow(
        <Input
          name="year"
          type="number"
          touched
          valid={false}
          messages={{ err: 'Error' }}
        />
      )
      expect(toJson(component)).toMatchSnapshot()
    })
  })
})
