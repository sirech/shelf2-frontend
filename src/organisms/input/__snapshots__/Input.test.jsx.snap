// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components Input displays error 1`] = `
<FormGroup
  tag="div"
>
  <Label
    for="year"
    tag="label"
    widths={
      Array [
        "xs",
        "sm",
        "md",
        "lg",
        "xl",
      ]
    }
  >
    year
  </Label>
  <DefaultConnectedControl
    component={[Function]}
    id="year"
    invalid={true}
    model=".year"
    name="year"
    type="number"
    valid={false}
  />
  <Modeled(Connect(Errors))
    component={[Function]}
    messages={
      Object {
        "err": "Error",
      }
    }
    model=".year"
    show={
      Object {
        "touched": true,
        "valid": false,
      }
    }
    wrapper={[Function]}
  />
</FormGroup>
`;

exports[`components Input renders correctly 1`] = `
<FormGroup
  tag="div"
>
  <Label
    for="title"
    tag="label"
    widths={
      Array [
        "xs",
        "sm",
        "md",
        "lg",
        "xl",
      ]
    }
  >
    title
  </Label>
  <DefaultConnectedControl
    component={[Function]}
    id="title"
    model=".title"
    name="title"
    type="text"
  />
  <Modeled(Connect(Errors))
    component={[Function]}
    model=".title"
    show={
      Object {
        "touched": true,
        "valid": false,
      }
    }
    wrapper={[Function]}
  />
</FormGroup>
`;

exports[`components Input renders correctly for different types 1`] = `
<FormGroup
  tag="div"
>
  <Label
    for="year"
    tag="label"
    widths={
      Array [
        "xs",
        "sm",
        "md",
        "lg",
        "xl",
      ]
    }
  >
    year
  </Label>
  <DefaultConnectedControl
    component={[Function]}
    id="year"
    model=".year"
    name="year"
    type="number"
  />
  <Modeled(Connect(Errors))
    component={[Function]}
    model=".year"
    show={
      Object {
        "touched": true,
        "valid": false,
      }
    }
    wrapper={[Function]}
  />
</FormGroup>
`;
